### PR TITLE
feat(chore/lint): add and configure stylelint mozaic plugin

### PR DESCRIPTION
**Pull request by mohamedMok**
## I have read [the contributing guidelines](https://mozaic.adeo.cloud/Contributing/)

- [X] Yes
- [ ] No

## Does this PR introduce a breaking change?

read [What is a breaking change ?](https://mozaic.adeo.cloud/Contributing/Prerequisite/GitConventions/#breaking-changes-)

- [] Yes
- [X] No

## Describe the changes

Github issue number or Jira issue URL: 
https://design-system-adeo.atlassian.net/browse/MZC-243

### CODE REVIEW

File: packages/css-dev-tools/styleLintConfig.js [hunk 1]
```diff
@@ -2,9 +2,19 @@
 
 module.exports = {
   extends: ['stylelint-config-standard'],
-  plugins: ['stylelint-scss'],
+  plugins: ['@mozaic-ds/stylelint-plugin-mozaic', 'stylelint-scss'],
   rules: {
     'at-rule-no-unknown': null,
-    'scss/at-rule-no-unknown': true,
+    'unit-whitelist': ['px', 'em', '%', 'rem', 'ms', 'deg', 'vw'],
```
Comment (tiloyi) [inline packages/css-dev-tools/styleLintConfig.js]: The "vh" unity is missing

File: packages/stylelint-plugin-mozaic/lib/splitSelector.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+// should return
+/* [
+    {
+        type: class,
+        subType: component,
+        string: 'cm-foo-bar__baz--foo',
+        bemStructure: [
+            { namespace: 'cm-' },
+            { block: 'foo-bar' },
+            { element: 'baz' },
```
Comment (tiloyi) [inline packages/stylelint-plugin-mozaic/lib/splitSelector.js]: The element can also be in kebab-case, I don't know if you took this case into account?

File: packages/stylelint-plugin-mozaic/test/index.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
+var testRule = require('stylelint-test-rule-tape')
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/test/index.js]: make a test for : `media: '/@',`
and for all other cases ('kebab-case', lowerCamelCase, snake_case, PascalCase)

File: packages/stylelint-plugin-mozaic/lib/checkCase.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+var stylelint = require('stylelint')
+
+const cases = {
+  'kebab-case': /^([a-z][a-z0-9]*)(-[a-z0-9]+)*$/,
+  lowerCamelCase: /^[a-z][a-zA-Z0-9]+$/,
+  snake_case: /^[a-z][a-zA-Z0-9]+$/,
+  PascalCase: /^[A-Z][a-zA-Z0-9]+$/,
+  UpperCamelCase: /^[A-Z][a-zA-Z0-9]+$/,
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/lib/checkCase.js]: no differences between pascal case and UpperCamelCase, the real name is pascal case.

File: packages/stylelint-plugin-mozaic/lib/splitSelector.js (hunk 2 above)
Comment (mohamedMok) [reply packages/stylelint-plugin-mozaic/lib/splitSelector.js]: You're right! Every part of the structure should be in kebab-case! 👍

File: packages/css-dev-tools/styleLintConfig.js (hunk 1 above)
Comment (gael-boyenval) [reply packages/css-dev-tools/styleLintConfig.js]: why this ? we are mostly listing all units here ? Can't we just allow everything ?

File: packages/stylelint-plugin-mozaic/index.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+}
+
+module.exports = stylelint.createPlugin(
+  ruleName,
+  (options, secondaryOptionObject) => {
+    const allOptions = { ...options, ...secondaryOptionObject }
+    return (root, result) => {
+      if (!allOptions) return
+      const validOptions = stylelint.utils.validateOptions(result, ruleName, {
+        wordDelimiterStyle: [isStringOrRegExp],
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/index.js]: can you check that it work with a regex, or disable the possibility ?

File: packages/stylelint-plugin-mozaic/index.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+        modifierDelimiter: [_.isString],
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/index.js]: why do we check only modifierDelimiter ? does this still reflect the config ? 

```
wordDelimiterStyle: 'kebab-case',
delimiters: {
   modifier: '--',
   element: '__',
   media: '/@',
},
```

Comment (gael-boyenval) [reply packages/stylelint-plugin-mozaic/index.js]: let's call `delimiters` => `bemEntitiesDelimiters` and `wordDelimiterStyle` => `caseStyle` ? what do you think ?

File: packages/css-dev-tools/styleLintConfig.js (hunk 1 above)
Comment (mohamedMok) [reply packages/css-dev-tools/styleLintConfig.js]: @gael-boyenval  We surely can just remove the rule to allow everything, it was mostly to test plugin rule with native ones.

File: packages/stylelint-plugin-mozaic/index.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      })
+      if (!validOptions) return
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/index.js]: We just return, without error handling ?

Comment (gael-boyenval) [reply packages/stylelint-plugin-mozaic/index.js]: not sure this all thing works

File: packages/stylelint-plugin-mozaic/lib/checkCase.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+        if (!isValidCase) {
+          isValid = false
+        }
+      })
+    }
+  })
+
+  if (!isValid) {
+    message = `Invalid class syntax: "${rule.selector}" should be "${options.wordDelimiterStyle}" (${ruleName})`
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/lib/checkCase.js]: we need to give a link to users to the doc related to the error. 
we should have a page listing all errors in the docs.

File: packages/css-dev-tools/styleLintConfig.js (hunk 1 above)
Comment (gael-boyenval) [reply packages/css-dev-tools/styleLintConfig.js]: ok

File: packages/stylelint-plugin-mozaic/index.js (hunk 7 above)
Comment (gael-boyenval) [reply packages/stylelint-plugin-mozaic/index.js]: we should test for bad user config also

File: packages/css-dev-tools/styleLintConfig.js [hunk 9]
```diff
@@ -5,15 +5,15 @@ module.exports = {
...
       {
-        wordDelimiterStyle: 'kebab-case',
-        delimiters: {
+        caseStyle: 'kebab-case',
+        bemEntitiesDelimiters: {
           modifier: '--',
           element: '__',
           media: '/@',
         },
+        prefixes: ['mc-', 'ml-', 'mu-', 'mt-'],
```
Comment (gael-boyenval) [inline packages/css-dev-tools/styleLintConfig.js]: add `mdu-` for `mozaic-dev-util`

File: packages/stylelint-plugin-mozaic/lib/splitSelector.js [hunk 10]
```diff
@@ -70,16 +69,23 @@ function splitBem(string, options) {
...
+  if (result.length === 1) return selector.replace(result[0], '')
+  if (result.length === 0) return selector
+}
+
 const splitSelector = (selector, options) => {
   selector.forEach((part, i) => {
     if (part.type === 'class') {
-      part.bemStructure = splitBem(part.value, options)
+      const value = removePrefixes(part.value, options)
+      part.bemStructure = splitBem(value, options)
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/lib/splitSelector.js]: we need to keep the prefix as part of the bem structure, not just remove it.

File: packages/stylelint-plugin-mozaic/lib/checkCase.js [hunk 11]
```diff
@@ -3,13 +3,15 @@ var stylelint = require('stylelint')
 const cases = {
   'kebab-case': /^([a-z][a-z0-9]*)(-[a-z0-9]+)*$/,
   lowerCamelCase: /^[a-z][a-zA-Z0-9]+$/,
```
Comment (gael-boyenval) [inline packages/stylelint-plugin-mozaic/lib/checkCase.js]: just `camelCase`

File: packages/css-dev-tools/styleLintConfig.js [hunk 12]
```diff
@@ -5,15 +5,15 @@ module.exports = {
...
-    'unit-whitelist': ['px', 'em', '%', 'rem', 'ms', 'deg', 'vw'],
     'plugin/mozaic-bem-pattern': [
       {
-        wordDelimiterStyle: 'kebab-case',
-        delimiters: {
+        caseStyle: 'kebab-case',
+        bemEntitiesDelimiters: {
           modifier: '--',
           element: '__',
           media: '/@',
```
Comment (gael-boyenval) [inline packages/css-dev-tools/styleLintConfig.js]: You are still not testing this : 
`cm-my-call--small\@from-m`

File: packages/stylelint-plugin-mozaic/lib/splitSelector.js (hunk 10 above)
Comment (gael-boyenval) [reply packages/stylelint-plugin-mozaic/lib/splitSelector.js]: something like this : 

```js
{
    identifier: 'mc-button',
    prefix: 'mc-',
    block: 'button',
    element: 'icon'
    modifier: 'large'
    media: 'from-m'
}
```

we test for `caseStyle` on `block`, `element`, `modifier` and `media`. 

we keep identifier as a kind of ID.

File: packages/css-dev-tools/styleLintConfig.js (hunk 12 above)
Comment (mohamedMok) [reply packages/css-dev-tools/styleLintConfig.js]: Good catch! It should be `media: '/@'`

File: packages/stylelint-plugin-mozaic/lib/splitSelector.js (hunk 10 above)
Comment (mohamedMok) [reply packages/stylelint-plugin-mozaic/lib/splitSelector.js]:
```
my-call--small@from-m
[
  { partType: 'block', string: 'my-call' },
  { partType: 'modifier', string: 'small' },
  { partType: 'media', string: 'from-m' },
  { partType: 'prefix', string: 'mc-' }
]
```

**Pull request merged**